File: python3/medium-solution/1081.py
"""
1081. Smallest Subsequence of Distinct Characters
    Given a string s, return the lexicographically smallest subsequence of s that contains all the distinct characters of s exactly once.

    Example :
    Input: s = "bcabc"
    Output: "abc"
"""



class Solution:
    def smallestSubsequence(self, s: str) -> str:
        freq = {c : 0 for c in s}
        visited = set()
        stack = []
        for c in s:
            freq[c] += 1

        for c in s:
            freq[c] -= 1
            if c in visited:
                continue
            while stack and stack[-1] > c and freq[stack[-1]] > 0:
                visited.remove(stack.pop())
            stack.append(c)
            visited.add(c)
        return "".join(stack)

File: JS/medium-solution/1081.js
/**
1081. Smallest Subsequence of Distinct Characters
    Given a string s, return the lexicographically smallest subsequence of s that contains all the distinct characters of s exactly once.

    Example :
    Input: s = "bcabc"
    Output: "abc"
 */



/**
 * @param {string} s
 * @return {string}
 */
var smallestSubsequence = function(s) {
    const freq = new Map();
    const visited = new Set();
    const stack = [];

    for (const c of s) {
        freq.set(c, (freq.get(c) || 0) + 1);
    }

    for (const c of s) {
        freq.set(c, freq.get(c) - 1);

        if (visited.has(c)) {
            continue;
        }

        while (
            stack.length > 0 &&
            stack[stack.length - 1] > c &&
            freq.get(stack[stack.length - 1]) > 0
        ) {
            visited.delete(stack.pop());
        }

        stack.push(c);
        visited.add(c);
    }

    return stack.join("");
};